File: lib/od/navigator/touch_navigator.py
from .base_navigator import BaseNavigator
from ...asset.directory import Directory
from ...asset.url import URL
from ...driver.support.driver_support import DriverSupport
import time


class TouchNavigator(BaseNavigator):
    """Navigator object that navigates purely through interaction"""

    def __init__(self, name, driver, opts, no_full_links=False):
        """Initializes Download Navigator object

        :param name: name of Navigator
        :param WebDriver driver: Selenium Webdriver
        :param Opts opts: Opts class
        :param bool no_full_links: Whether or not file links can be recorded from this OD
        """
        super().__init__(name, driver, opts, no_full_links)
        self._cur_dir = Directory(0, URL(self._driver.current_url))

    def _get_nav_info_elements(self, nav_info) -> list:
        return DriverSupport.get_elements_all(self._driver, self._opts, nav_info.css_name,
                                              nav_info.wait_err_message)

    def _go_to_directory(self, directory) -> None:
        self._move_up_to_destination(directory)
        self._move_down_to_destination(directory)

    def _move_up_to_destination(self, directory) -> None:
        """Move up the directory tree before heading down to the destination directory

        :param Directory directory: Destination directory
        :return:
        """
        while self._cur_dir.url not in directory.url and self._cur_dir.url != \
                self._driver.current_url:
            time.sleep(3)
            is_home = self._go_back_button()
            if is_home:
                self._cur_dir = Directory(0, self._driver.current_url)
                break
            self._cur_dir.pop_path()
            self._cur_dir.depth_level -= 1

    def _move_down_to_destination(self, directory) -> None:
        """Move down the directory tree towards destination directory

        :param Directory directory: Destination Directory
        :return:
        """
        while self._cur_dir.depth_level != directory.depth_level:
            time.sleep(3)
            elements = DriverSupport.get_elements_all(self._driver, self._opts,
                                                      self._scraper.nav_info.css_select,
                                                      self._scraper.nav_info.wait_err_message)
            for el in elements:
                name_element = DriverSupport.get_element(el, self._scraper.nav_info.css_name)
                name = name_element.text.strip()
                name = self._scraper.apply_filter(name)
                if name in directory.url:
                    el.click()
                    self._cur_dir.depth_level += 1
                    break
        self._cur_dir = directory

    def _go_back_button(self) -> bool:
        """Interact with the OD to go back a page"""
        pass

File: app/src/main/java/xyz/zimtools/zyod/od/navigators/TouchNavigator.java
package xyz.zimtools.zyod.od.navigators;

import org.openqa.selenium.By;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.remote.RemoteWebDriver;
import xyz.zimtools.zyod.AppConfig;
import xyz.zimtools.zyod.args.Args;
import xyz.zimtools.zyod.assets.Directory;
import xyz.zimtools.zyod.assets.ODUrl;
import xyz.zimtools.zyod.od.ODType;
import xyz.zimtools.zyod.support.NavSupport;

import java.util.List;
import java.util.Map;

/**
 * Navigate an OD purely through simulated interaction
 */
abstract class TouchNavigator extends NameNavigator {
    protected Directory curDir;

    public TouchNavigator(ODType id, RemoteWebDriver driver, Args args) {
        super(id, driver, args);
        this.curDir = new Directory(0, new ODUrl(driver.getCurrentUrl()));
    }

    @Override
    protected Map<String, Object> goToDirectory(Directory directory) {
        this.moveUpToDestination(directory);
        this.moveDownToDestination(directory);
        return Map.of(NOT_LOGIN_KEY, true, DIRECTORY_KEY, directory);
    }

    /**
     * Move up the directory tree before heading down to the destination directory
     *
     * @param destDir Destination {@link Directory}
     */
    protected void moveUpToDestination(Directory destDir) {
        while (!destDir.toString().contains(this.curDir.toString()) &&
                !this.curDir.toString().equals(this.driver.getCurrentUrl())) {
            long interactWait = this.args.getArgsInteractive().getInteractWait();
            AppConfig.sleep(interactWait);
            boolean isHome = this.goBackButton();
            if (isHome) {
                this.curDir = new Directory(0, new ODUrl(this.driver.getCurrentUrl()));
                break;
            }
            this.curDir.popPath();
        }
    }

    /**
     * Move down the directory tree to destination directory
     *
     * @param destDir destination {@link Directory}
     */
    protected void moveDownToDestination(Directory destDir) {
        while (this.curDir.getDepthLevel() != destDir.getDepthLevel()) {
            long interactWait = this.args.getArgsInteractive().getInteractWait();
            AppConfig.sleep(interactWait);
            List<WebElement> elements = NavSupport.getElements(this.driver,
                    this.scraper.getNavInfo().getCssFileSelector());
            for (WebElement el : elements) {
                WebElement nameElement =
                        el.findElement(By.cssSelector(this.scraper.getNavInfo().getCssFileName()));
                String fileName = nameElement.getText().strip();
                fileName = this.scraper.applyFilter(fileName);
                if (destDir.toString().contains(fileName)) {
                    el.click();
                    break;
                }
            }
            this.curDir = destDir;
        }
    }

    /**
     * Interaction with the OD to go back a page
     *
     * @return True if back at home page; false otherwise
     */
    protected abstract boolean goBackButton();
}